File: ASSIGNMENT10/c/Q.7.c
#include <stdio.h>
#include <stdlib.h>

struct TNode {
    int data;
    struct TNode *left;
    struct TNode *right;
    int lthread, rthread; // lthread and rthread indicate if left/right is a thread
};

struct TNode* createThreadedNode(int data) {
    struct TNode* newNode = (struct TNode*)malloc(sizeof(struct TNode));
    newNode->data = data;
    newNode->left = newNode->right = NULL;
    newNode->lthread = newNode->rthread = 0;
    return newNode;
}

void insert(struct TNode** root, int data) {
    struct TNode* newNode = createThreadedNode(data);
    if (*root == NULL) {
        *root = newNode;
        return;
    }

    struct TNode* current = *root;
    struct TNode* parent = NULL;

    while (current != NULL) {
        parent = current;
        if (data < current->data) {
            if (current->lthread == 0)
                current = current->left;
            else
                break;
        } else {
            if (current->rthread == 0)
                current = current->right;
            else
                break;
        }
    }

    if (data < parent->data) {
        parent->left = newNode;
        parent->lthread = 0; // left is not a thread anymore
        newNode->left = parent->left; // thread to in-order predecessor
        newNode->right = parent; // thread to parent
        newNode->rthread = 1; // right is a thread
    } else {
        parent->right = newNode;
        parent->rthread = 0; // right is not a thread anymore
        newNode->right = parent->right; // thread to in-order successor
        newNode->left = parent; // thread to parent
        newNode->lthread = 1; // left is a thread
    }
}

void inorder(struct TNode* root) {
    if (root == NULL) return;

    struct TNode* current = root;
    while (current->lthread == 0) {
        current = current->left;
    }

    while (current != NULL) {
        printf("%d ", current->data);
        if (current->rthread == 1) {
            current = current->right;
        } else {
            current = current->right;
            while (current != NULL && current->lthread == 0) {
                current = current->left;
            }
        }
    }
}

int main() {
    struct TNode* root = NULL;
    insert(&root, 20);
    insert(&root, 10);
    insert(&root, 30);
    insert(&root, 5);
    insert(&root, 15);

    printf("Inorder traversal of threaded binary tree: ");
    inorder(root);
    printf("\n");
    return 0;
}

File: ASSIGNMENT10/python/Q.7.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None
        self.is_threaded = False  # Indicates if the right pointer is a thread

class ThreadedBinaryTree:
    def __init__(self):
        self.root = None

    def insert(self, data):
        new_node = Node(data)
        if self.root is None:
            self.root = new_node
            return
        
        current = self.root
        parent = None
        
        while current:
            parent = current
            if data < current.data:
                current = current.left
            else:
                current = current.right

        if data < parent.data:
            parent.left = new_node
            # Create a thread to the parent
            new_node.right = parent
            new_node.is_threaded = True
        else:
            parent.right = new_node
            # Create a thread to the parent
            new_node.right = self.inorder_successor(parent)
            if new_node.right:
                new_node.is_threaded = True

    def inorder_successor(self, node):
        current = node.right
        while current and current.left:
            current = current.left
        return current

    def inorder_traversal(self):
        current = self.root
        # Go to the leftmost node
        while current and current.left:
            current = current.left
        
        # Now current is the leftmost node
        while current:
            print(current.data, end=' ')
            # If the node has a thread, use it
            if current.is_threaded:
                current = current.right
            else:
                current = current.right
                # Move to the leftmost node of the right subtree
                while current and current.left:
                    current = current.left

# Example Usage
if __name__ == "__main__":
    tbt = ThreadedBinaryTree()
    data = [20, 10, 30, 5, 15, 25, 35]
    
    for value in data:
        tbt.insert(value)

    print("In-order Traversal of the Two-Way Threaded Binary Tree:")
    tbt.inorder_traversal()
